lib/Corres_Cases.thy

```diff
@@ -0,0 +1,319 @@
+
+  val typeNm        = (String.concatWith "." o rev o tl o rev) constNm_fds;
+  val split         = Proof_Context.get_thm ctxt (typeNm ^ ".split");
+  val vars = Term.add_vars (Thm.prop_of split) []
+
+  val datatype_name = List.nth (rev constNm_fds,1)
+
+  fun T_is_datatype (Type (nm,_)) = (Long_Name.base_name nm) = (Long_Name.base_name datatype_name)
+    | T_is_datatype _ = false
+
+  val datatype_var =
+    case (find_first (fn ((_,_),T') => (T_is_datatype T')) vars) of
+      SOME (ix,_) => ix
+    | NONE => error ("Couldn't find datatype in thm: " ^ datatype_name)
+
+  val split' = Drule.infer_instantiate ctxt
+    [(datatype_var, Thm.cterm_of ctxt (List.last args))] split
+
+in
+   SOME split' end
+   handle TERM _ => NONE;
+\<close>
+
+(* The above function as an attribute. The term argument is expected to be a case statement. *)
+attribute_setup get_split_rule = \<open>Args.term >>
+  (fn t => Thm.rule_attribute [] (fn context => fn _ =>
+      case (get_split_rule (Context.proof_of context) t) of
+        SOME thm => thm
+      | NONE => Drule.free_dummy_thm))\<close>
+
+(* Apply a split rule to a goal. Example usage:
+
+   apply_split f "\<lambda>f. corres_underlying sr nf nf' r P P' f f'"
+
+   The corres_underlying term provides the context, and f is the term we expect to be
+   a case statement.  *)
+method apply_split for f :: 'a and R :: "'a \<Rightarrow> bool" =
+    (match [[get_split_rule f]] in U: "(?x :: bool) = ?y" \<Rightarrow>
+      \<open>match U[THEN iffD2] in U': "\<And>H. ?A \<Longrightarrow> H (?z :: 'c)" \<Rightarrow>
+        \<open>match (R) in "R' :: 'c \<Rightarrow> bool" for R' \<Rightarrow>
+          \<open>rule U'[where H=R']\<close>\<close>\<close>)
+
+context
+begin
+
+(* This predicate provides an abstraction for guard/precondition terms for transformations
+   on those guards.
+
+   P and P' are the abstract and concrete preconditions before the transformation
+   Q and Q' are the abstract and concrete preconditions after the transformation
+
+   R is the predicate to be transformed.
+*)
+private definition corres_case_helper ::
+  "(('a \<Rightarrow> bool) \<times> ('b \<Rightarrow> bool)) \<Rightarrow> (('a \<Rightarrow> bool) \<times> ('b \<Rightarrow> bool)) \<Rightarrow> bool \<Rightarrow> bool" where
+  "corres_case_helper \<equiv> \<lambda>(P, P') (Q, Q') R. (\<forall>s. P s \<longrightarrow> Q s) \<longrightarrow>  (\<forall>s. P' s \<longrightarrow> Q' s) \<longrightarrow> R"
+
+
+(* The following lemmas enable us to lift preconditions of corres_case_helper over conjunction,
+   universal quantifiers, and implication. Note that there are strong versions for forall/implies
+   where both guards are treated uniformly, and weak versions, where forall/implies is dropped
+   in one guard, but not the other.
+*)
+
+private lemma corres_case_helperI:
+  "corres_case_helper (P, P') (P, P') R \<Longrightarrow> R"
+  by (simp add: corres_case_helper_def)
+
+private lemma corres_case_conj_process:
```

Comment:
I'm not clear why these are all named `_process` ... is that because they're meant to be used for processing of the ... er, I'm not sure what.

Reply:
Ah, R, the term to be transformed?

Reply:
Would `_lift` make sense, given the description above says `enable us to lift`?

Reply:
The terminology is from `wpc`, which has almost the same lemmas, and calls them `wpc_helper_X_process`. `_lift` would probably be a better name, but then that connection is lost. I could add a comment that this is where the names come from.

Reply:
Would help, yes, thank you